documentation/proposals/Proposal - Axis Input Devices.md

```diff
@@ -0,0 +1,642 @@
+Due to the complex set of constraints detailed below, it is important that a validation layer exists to ensure that higher-level code can trust the device is following important conventions and constraints. As a result, each `IAxisDevice` must be validated, similar to Vulkan's validation layers, before `InputContext` can use the device to create a wrapper/implementation of the higher-level interfaces defined in Multi-Backend Input (or otherwise make use of the raw axes). This validation is done by `InputContext` if used, however if the user is using their own device aggregator type then it is important that the validation implementation is easily callable by those implementations due to its inevitable complexity.
+
+This validation shall be exposed as follows:
+
+```csharp
+namespace Silk.NET.Input;
+public partial interface IAxisDevice
+{
+    static sealed void Validate(IAxisDevice device);
+}
+```
+
+
+## IAxisDevice
+
+```csharp
+/// <summary>
+/// This represents a single input device, physical or virtual, that is comprised of a collection of axes (or floats)
+/// The IReadOnlyList<float> interface is used to provide the normalized values of each axis, if it is indeed normalized (most should be). Otherwise, it would provide the raw, unnormalized value.
+/// </summary>
+public interface IAxisDevice : IReadOnlyList<float>, IInputDevice
+{
+    /// <summary>
+    /// The main entry-point for useful information about common axis groups - e.g. joysticks, trackpads, buttons, gyros, etc.
+    /// Turns single axes/floats into more broadly useable information
+    /// Must be provided in order of <see cref="AxisGroup.Index"/>
+    /// </summary>
+    IReadOnlyList<AxisGroup> Groups { get; }
+
+    /// <summary>
+    /// The description of each axis in the hardware device
+    /// This information is used primarily to allow for programmatically defined axis groups, axis devices, runtime validation,
+    /// and custom device mappings and handling.
+    /// Must be provided in order of <see cref="AxisDescription.Index"/>
+    /// </summary>
+    IReadOnlyList<AxisDescription> Axes { get; }
+
+    /// <summary>
+    /// Provides raw values for each axis as defined by the device's native API
+    /// Must be provided in order of <see cref="AxisDescription.Index"/>
+    /// </summary>
+    IReadOnlyList<float> RawValues { get; }
+}
+```
+
+## IAxisInputHandler
+
+This is how a developer could subscribe directly to the axis changes of an `IAxisDevice`. It mimicks the `IGamepadInputHandler` interface from the [Multi-Backend Input proposal](./Proposal - Multi-Backend Input.md).
+
+```csharp
+
+/// <summary>
+/// Represents the state of a single axis
+/// </summary>
+/// <param name="IsActive">True if the axis is currently in use / actively updated (required to be true if axis is not <see cref="AxisTrait.Dynamic"/>)
+public readonly record struct AxisState(IAxisDevice Device, AxisDescription Description, float Value, bool IsActive = true);

```

Comment:
Should `IsActive` be in `AxisDescription`? Or more specifically, should this be queryable outside of a handler method call?

Reply:
I like the idea of putting it in `AxisDescription`, though I'd really like to avoid the scenario where a "fixed" axis is marked as "IsActive = false" post-validation. currently theres no defense against this anyway, but it could create ideas.

though it could be a nice way to say "hey my X button is broken, can we not?" so in that case im actually gonna update this to be in the AxisDefinition and allow IsActive (w/ potential rename) to be used universally. we're probably never going to do that, but ¯\_(ツ)_/¯